python/lsst/daf/butler/datastores/posixDatastore.py

```diff
@@ -168,9 +174,31 @@ def put(self, inMemoryDataset, storageClass, storageHint, typeName=None):
                     if compTypeName is not None:
                         compTypeName = "{}.{}".format(compTypeName, comp)
                     compUris[comp], _ = self.put(info.component, info.storageClass,
-                                                 location.componentUri(comp), compTypeName)
+                                                 dataUnits, compTypeName)
                 return None, compUris
 
+        # Get a location from the templates
+        template = None
+        component = None
+        if typeName is not None:
+            if typeName in self.templates:
+                template = self.templates[typeName]
+            elif "." in typeName:
+                baseType, component = typeName.split(".", maxsplit=1)
+                if baseType in self.templates:
+                    template = self.templates[baseType]
+
+        if template is None:
+            if "default" in self.templates:
+                template = self.templates["default"]
```

Comment:
This loading could go into a separate function or `classmethod` of `Template`.

Reply:
Yes. Ok, I would seed it with the config object.

Reply:
This is determining the right template to use based on the dataset type by looking up in the datastore's template registry. I think I can make this work like the Factory we use for formatters and storage classes (although we are storing instances and not classes).